openghg/objectstore/metastore/_classic_metastore.py

```diff
@@ -0,0 +1,159 @@
+"""
+This module implements a MetaStore based on TinyDB, with custom
+middleware based on `CachingMiddleware`.
+
+The metastore can be accessed given a bucket and data type.
+ In a context manager, use `open_metastore`. Outside of a context
+ manager, use `ClassicMetaStore.from_bucket`; in this case, the
+metastore must be closed using the `close` method.
+
+Closing the metastore is necessary, since `CachingMiddleware`
+doesn't write to disk unless at least 1000 writes have been made.
+
+The `ClassicMetaStore` is organised in the same way that the
+metastore was organised in OpenGHG <= v 6.2: there are separate
+TinyDB databases for each data type.
+
+Besides taking a data type parameter in the `__init__` method,
+`ClassicMetaStore` only extends `TinyDBMetastore` by adding
+a `from_bucket` class method, and a `close` method.
+These methods are only needed by `store.BaseStore`.
+
+"""
+from __future__ import annotations
+
+from collections.abc import Generator
+from contextlib import contextmanager
+import json
+from typing import Literal, Optional, TypeVar
+
+import tinydb
+from tinydb.middlewares import CachingMiddleware
+
+from openghg.objectstore import exists, get_object, set_object_from_json
+from openghg.objectstore.metastore._metastore import TinyDBMetaStore
+from openghg.types import MetastoreError
+
+
+object_store_data_classes = {
```

Comment:
I think openghg.store.spec needs to be changed because it creates circular dependencies (e.g. with objectstore._integrity), which are hidden by importing within functions instead of at top level.

Ideally there would be a config subpackage that configures object stores, and this info should probably be read from a config file. After the next PR in the sequence, it might be the case that only things within objectstore need to know about these keys. 

One solution is to put store.spec outside of store for now, or to wait until the next PR in the sequence to see if these don't need to be outside objectstore.

A better solution would be to have a different tables for each data type inside a single metastore. This is supported by TinyDB and MongoDB, but I didn't want to do it now because it's too much of a change from the current set up.

Reply:
I didn't resolve this, but I added a TODO comment to move this code in the next iteration (Part 4, related to issue #727 ). I suspect that after `BaseStore` is refactored, the only place this info will be needed is in the `objectstore` module. (If I do it correctly, this should be the case, since it is info internal to the object store.)

Reply:
> I think openghg.store.spec needs to be changed because it creates circular dependencies (e.g. with objectstore._integrity), which are hidden by importing within functions instead of at top level.

Ok - one comment on the circular imports is that I did find with careful ordering of imports and only importing exactly where needed I was able to resolve these when they came up in code.

> One solution is to put store.spec outside of store for now, or to wait until the next PR in the sequence to see if these don't need to be outside objectstore.

Yes it would be fine for the details to be in a separate sub-package if needed

> A better solution would be to have a different tables for each data type inside a single metastore. This is supported by TinyDB and MongoDB, but I didn't want to do it now because it's too much of a change from the current set up.

Yes, I think we would need to discuss if that was the best solution so we don't want to currently make this change (at least at the moment)

> I didn't resolve this, but I added a TODO comment to move this code in the next iteration (Part 4, related to issue https://github.com/openghg/openghg/issues/727 ). I suspect that after BaseStore is refactored, the only place this info will be needed is in the objectstore module. (If I do it correctly, this should be the case, since it is info internal to the object store.)

Ok that's fine as long as this is tracked. Would be good to add this as a comment on Issue #766 so it's clear this still needs to be done.

Reply:
I referenced this discussion on Issue #727. (I think Issue #766 can be closed with this PR, since 766 is just splitting off the metastore from `BaseStore`. I guess "part 4" is a misnomer, since the next PR in the sequence will address #727.)